core/circuit.go

```diff
@@ -0,0 +1,247 @@
+type Circuit struct {
+	log    *util.Logger
+	uiChan chan<- util.Param
+
+	maxCurrent    float64           // max allowed current. will not be used when 0
+	maxPower      float64           // max allowed power. will not be used when 0
+	parentCircuit *Circuit          // parent circuit reference, used to determine current limits from hierarchy
+	phaseCurrents api.PhaseCurrents // meter to determine phase current
+	powerMeter    api.Meter         // meter to determine current power
+}
+
+// NewCircuitFromConfig creates a new Circuit
+func NewCircuitFromConfig(log *util.Logger, circuits map[string]*Circuit, vMeters map[string]*VMeter, other map[string]interface{}) (*Circuit, *VMeter, string, error) {
+	var cc struct {
+		Name       string  `mapstructure:"name"`       // unique name, used as reference in lp
+		MaxCurrent float64 `mapstructure:"maxCurrent"` // the max allowed current of this circuit
+		MaxPower   float64 `mapstructure:"maxPower"`   // the max allowed power of this circuit (kW)
+		MeterRef   string  `mapstructure:"meter"`      // Charge meter reference
+		ParentRef  string  `mapstructure:"parent"`     // name of parent circuit
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, nil, "", err
+	}
+
+	// TODO circuits as all other entities MUST NOT know their name
+	if cc.Name == "" {
+		return nil, nil, "", fmt.Errorf("failed configuring circuit, need to have a name")
+	}
+
+	if _, ok := circuits[cc.Name]; ok {
+		return nil, nil, "", fmt.Errorf("circuit name already used: %s", cc.Name)
+	}
+
+	var parent *Circuit
+	if cc.ParentRef != "" {
+		var ok bool
+		if parent, ok = circuits[cc.ParentRef]; !ok {
+			return nil, nil, "", fmt.Errorf("parent circuit not defined: %s", cc.ParentRef)
+		}
+	}
+
+	var vMeter *VMeter
+	var meterPhaseCurrents api.PhaseCurrents
+	var meterPower api.Meter
+
+	// check meter exists
+	if cc.MeterRef != "" {
+		dev, err := config.Meters().ByName(cc.MeterRef)
+		if err != nil {
+			return nil, nil, "", err
+		}
+		var m api.Meter = dev.Instance()
+
+		meterPower = m
+
+		var ok bool
+		if meterPhaseCurrents, ok = m.(api.PhaseCurrents); !ok {
+			if cc.MaxCurrent != 0 {
+				return nil, nil, "", fmt.Errorf("meter %s does not support phase currents, but current checking configured", cc.MeterRef)
+			}
+		}
+	} else {
+		// no dedicated vMeter given, create vmeter
+		vMeter = NewVMeter(cc.Name)
+		meterPhaseCurrents = vMeter
+		meterPower = vMeter
+	}
+
+	newCC, err := NewCircuit(log, cc.MaxCurrent, cc.MaxPower*1000, parent, meterPhaseCurrents, meterPower)
+
+	// if we have parent circuit and this one uses a vMeter, add this circuit as consumer
+	if cc.ParentRef != "" {
+		if vm, ok := vMeters[cc.ParentRef]; ok {
+			vm.AddConsumer(newCC)
+		}
+	}
+	return newCC, vMeter, cc.Name, err
+}
+
+// NewCircuit a circuit with defaults
+func NewCircuit(log *util.Logger, limitCurrent float64, limitPower float64, p *Circuit, pc api.PhaseCurrents, pm api.Meter) (*Circuit, error) {
```

Comment:
Might be nice to have an `api.Circuit` interface in addition to the concrete type.